src/constraints.jl

```diff
@@ -0,0 +1,533 @@
+    @assert JuMP.is_valid(model, cref) "Disjunctive constraint does not belong to model."
+    cidx = JuMP.index(cref)
+    dict = _disjunctions(model)
+    #delete each logical variable
+    for lv_idx in dict[cidx].constraint.disjuncts
+        JuMP.delete(model, LogicalVariableRef(model, lv_idx))
+    end
+    #delete from gdp_data
+    delete!(dict, cidx)
+    #not ready to optimize
+    _set_ready_to_optimize(model, false)
+    return 
+end
+function JuMP.delete(model::JuMP.Model, cref::DisjunctConstraintRef)
+    @assert JuMP.is_valid(model, cref) "Disjunctive constraint does not belong to model."
+    cidx = JuMP.index(cref)
+    dict = _disjunct_constraints(model)
+    #delete from gdp_data
+    delete!(dict, cidx)
+    #delete from constraint_to_indicator mapping
+    delete!(_constraint_to_indicator(model), cidx)
+    # TODO: delete reformulated constraint? This would require mapping cidx to reformulated constraint index
+    #not ready to optimize
+    _set_ready_to_optimize(model, false)
+    return 
+end
+function JuMP.delete(model::JuMP.Model, cref::LogicalConstraintRef)
+    @assert JuMP.is_valid(model, cref) "Logical constraint does not belong to model."
+    cidx = JuMP.index(cref)
+    dict = _logical_constraints(model)
+    #delete from gdp_data
+    delete!(dict, cidx)
+    #not ready to optimize
+    _set_ready_to_optimize(model, false)
+    return 
+end
+
+################################################################################
+#                              Disjunct Constraints
+################################################################################
+"""
+    JuMP.build_constraint(
+        _error::Function, 
+        func, 
+        set::_MOI.AbstractScalarSet,
+        tag::DisjunctConstraint
+    )::_DisjunctConstraint
+
+Extend `JuMP.build_constraint` to add constraints to disjuncts. This in 
+combination with `JuMP.add_constraint` enables the use of 
+`@constraint(model, [name], constr_expr, tag)`, where tag is a
+`DisjunctConstraint(::Type{LogicalVariableRef})`. The user must specify the 
+`LogicalVariable` to use as the indicator for the `_DisjunctConstraint` being created.
+"""
+function JuMP.build_constraint(
+    _error::Function, 
+    func, 
+    set::_MOI.AbstractScalarSet, 
+    tag::DisjunctConstraint
+)
+    constr = JuMP.build_constraint(_error, func, set)
+    return _DisjunctConstraint(constr, tag.indicator)
+end
+
+# Allows for building DisjunctConstraints for VectorConstraints since these get parsed differently by JuMP (the set is changed to a MOI.AbstractScalarSet)
+function JuMP.build_constraint(
+    _error::Function, 
+    func::Vector{T}, 
+    set::Union{_MOI.LessThan{Bool}, _MOI.GreaterThan{Bool}, _MOI.EqualTo{Bool}},
+    tag::DisjunctConstraint
+) where {T <: JuMP.AbstractJuMPScalar}
+    dim = length(func)
+    mapped_set = _scalar_to_vec_set(set,dim)
+    constr = JuMP.build_constraint(_error, func, mapped_set)
+    return _DisjunctConstraint(constr, tag.indicator)
+end
```

Comment:
This approach will likely have unintended behavior and likely to break with future JuMP feature additions. It would be simpler and more stable to instead define:
```julia
function JuMP.build_constraint(
    _error::Function, 
    func, 
    set::JuMP.Nonnegatives, 
    tag::DisjunctConstraint
)
    constr = JuMP.build_constraint(_error, func, set)
    return _DisjunctConstraint(constr, tag.indicator)
end
```
for each of the three JuMP sets `Nonnegatives`, `Nonpositives`, and `Zeros`.

Reply:
I agree, but the problem is that if we do this, the method is never called because these constraints will first go through JuMP's method:
```
function build_constraint(
    _error::Function,
    f,
    set::Nonpositives,
    args...;
    kwargs...,
)
    return build_constraint(_error, f, MOI.LessThan(false), args...; kwargs...)
end
```

Reply:
nevermind, your suggestion works